tests/core/test_nlg.py

```diff
@@ -328,3 +331,271 @@ def test_nlg_fill_response_quick_replies(
+
+@pytest.mark.parametrize(
+    ("slot_name", "slot_value", "output_channel", "response_variation"),
+    (
+        ("slot test", "A", "os", "Conditional Response A"),
+        ("slot test", "A", "app", "Conditional app response A"),
+        ("slot test", "B", "", "Conditional Response B"),
+        ("slot test", "C", "", "Default response"),
+    ),
+)
+async def test_nlg_conditional_response_variations(
+    slot_name, slot_value, output_channel, response_variation
+):
+    responses = {
+        "utter_test": [
+            {
+                "text": "Conditional Response A",
+                "condition": [{"type": "slot", "name": "slot test", "value": "A"}],
+                "channel": "os",
+            },
+            {
+                "text": "Conditional app response A",
+                "condition": [{"type": "slot", "name": "slot test", "value": "A"}],
+                "channel": "app",
+            },
+            {
+                "text": "Conditional Response B",
+                "condition": [{"type": "slot", "name": "slot test", "value": "B"}],
+            },
+            {"text": "Default response"},
+        ]
+    }
+    t = TemplatedNaturalLanguageGenerator(responses=responses)
+
+    slot = Slot(name=slot_name, initial_value=slot_value, influence_conversation=False)
+    tracker = DialogueStateTracker(sender_id="nlg_test", slots=[slot])
+
+    r = await t.generate(
+        utter_action="utter_test", tracker=tracker, output_channel=output_channel
+    )
+    assert r.get("text") == response_variation
+
+    no_slots_tracker = DialogueStateTracker(sender_id="nlg_test_default", slots=None)
+    default_response = await t.generate(
+        utter_action="utter_test", tracker=no_slots_tracker, output_channel=""
+    )
+    assert default_response.get("text") == "Default response"
+
+
+async def test_nlg_when_multiple_conditions_satisfied():
+    responses = {
+        "utter_action": [
+            {
+                "text": "example A",
+                "condition": [{"type": "slot", "name": "test", "value": "A"}],
+            },
+            {
+                "text": "example B",
+                "condition": [{"type": "slot", "name": "test_another", "value": "B"}],
+            },
+        ]
+    }
+
+    t = TemplatedNaturalLanguageGenerator(responses=responses)
+    slot_a = Slot(name="test", initial_value="A", influence_conversation=False)
+    slot_b = Slot(name="test_another", initial_value="B", influence_conversation=False)
+    tracker = DialogueStateTracker(sender_id="test_nlg", slots=[slot_a, slot_b])
+    resp = await t.generate(
+        utter_action="utter_action", tracker=tracker, output_channel=""
+    )
+    assert resp.get("text") in ["example A", "example B"]
+
+
+@pytest.mark.parametrize(
+    (
+        "first_slot",
+        "first_slot_value",
+        "second_slot",
+        "second_slot_value",
+        "response_variation",
+    ),
+    (
+        ("test", "A", "test_other", "B", "example AB"),
```

Comment:
how about a test where a slot is `null`?

Reply:
or where there is a slot which isn't referenced by any constraint?

Reply:
and a test for a condition which requires a slot to be `None` would be interesting!

Reply:
The first two suggestions are very good and prove that we need to enforce a default for these cases. But I'll see what errors are raised when no response variation is specified as default and slot is null.
I'm not entirely sure about the third suggestion - is that in the case when a user specifies None value for a slot in `domain.yml` ?

Reply:
When the slot is null or not referenced by any constraint, and no response variation as default, then the response generated is None - this could cause AttributeError when later the assistant tries to get value for `text` key. I'll list this as something we should recommend in the docs - to cover all bases with a default.

Reply:
> I'm not entirely sure about the third suggestion - is that in the case when a user specifies None value for a slot in domain.yml ?

Let's say there is a slot `name` and for some reason the user doesn't provide it which means it becomes `None` (it could have had a different value before).

Is this

```yaml
responses:
  utter_hi:
  - text: "Hello"
```

the same as this?

```yaml
responses:
  utter_hi:
  - text: "Hello"
     condition:
     - type: slot
        name: name
        value: null
```

would in my opinion still be nice to have a test to ensure this is the same / works, what do you think?